File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    let answer = 0
    const left = new Array(topping.length).fill(0)
    const right = new Array(topping.length).fill(0)
    
    const leftSet = new Set()
    const rightSet = new Set()
    
    for(let i = 0; i< topping.length; i++){
        leftSet.add(topping[i]);
        left[i] = leftSet.size;
    }
    
    for(let i = topping.length - 1; i >= 0; i--){
        rightSet.add(topping[i]);
        right[i] = rightSet.size;
    }
    console.log(left, right)
    for(let i = 0; i < topping.length-1; i++){
        if(left[i] === right[i + 1]){
            answer++
        }
    }
    return answer
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    answer = 0
    n = len(topping)
    left = [0] * len(topping)
    right = [0] * len(topping)
    left_set = set()
    right_set = set()
    
    for i in range(n):
        left_set.add(topping[i])
        left[i] = len(left_set)
    for i in range(n-1, -1, -1):
        right_set.add(topping[i])
        right[i] = len(right_set)
    
        
    for i in range(0, n-1):
        if left[i] == right[i+1]:
            answer+=1
    return answer